File: packages/depz-sensor-sdk-python/depz_sensor_sdk/protocol/sr04.py
"""SR04 wire codecs (contracts/03_SENSOR_SR04.md)."""

from __future__ import annotations

import struct
from dataclasses import dataclass
from enum import IntEnum


class Sr04Cmd(IntEnum):
    GET_SAMPLE_PERIOD = 0x32
    SET_SAMPLE_PERIOD = 0x33
    GET_ECHO_DECAY = 0x34
    SET_ECHO_DECAY = 0x35
    MEASURE_ONCE = 0x36
    START_MEASUREMENT_LOOP = 0x37
    STOP_MEASUREMENT_LOOP = 0x38


class Sr04Rpt(IntEnum):
    DATA = 0x91
    SAMPLE_PERIOD = 0x92
    ECHO_DECAY = 0x93


ECHO_TIMEOUT = 0xFFFF  # echo_time_us sentinel: no echo received

SAMPLE_PERIOD_DEFAULT_US = 50_000
ECHO_DECAY_DEFAULT_US = 5_000
#: Window the *device* silently clamps SET_ECHO_DECAY into (contract 03 §3).
ECHO_DECAY_MIN_US = 4_000
ECHO_DECAY_MAX_US = 65_000
#: Widest value each wire field can carry. Distinct from the clamp window
#: above: sending 0xFFFF is legal (the device clamps it), sending 0x10000 is
#: not representable at all. `sr04.py` validates against these so an
#: out-of-range argument raises ValueError instead of leaking struct.error.
ECHO_DECAY_WIRE_MAX_US = 0xFFFF  # u16
SAMPLE_PERIOD_WIRE_MAX_US = 0xFFFF_FFFF  # u32


@dataclass(frozen=True)
class Sr04Data:
    source_cmd: int  # 0x36 single shot (host or SYNC_IN), 0x37 loop sample
    timestamp_us: int
    echo_time_us: int

    @classmethod
    def unpack(cls, payload: bytes) -> "Sr04Data":
        cmd, ts, echo = struct.unpack("<BQH", payload)
        return cls(cmd, ts, echo)


def pack_sample_period(period_us: int) -> bytes:
    return struct.pack("<I", period_us)


def unpack_sample_period(payload: bytes) -> int:
    return struct.unpack("<I", payload)[0]


def pack_echo_decay(decay_us: int) -> bytes:
    return struct.pack("<H", decay_us)


def unpack_echo_decay(payload: bytes) -> int:
    return struct.unpack("<H", payload)[0]


def distance_mm_from_echo(echo_time_us: int, air_temp_c: float | None = None) -> float | None:
    """Round-trip echo time → distance in mm; None for the timeout sentinel.

    Default speed of sound 343 m/s; with `air_temp_c` uses
    c = 331.3 + 0.606·T (m/s).
    """
    if echo_time_us == ECHO_TIMEOUT:
        return None
    c_m_s = 343.0 if air_temp_c is None else 331.3 + 0.606 * air_temp_c
    return echo_time_us * c_m_s / 2000.0

File: packages/depz-sensor-sdk-ts/src/protocol/sr04.ts
/** SR04 wire codecs (contracts/03_SENSOR_SR04.md). */

export enum Sr04Cmd {
  GetSamplePeriod = 0x32,
  SetSamplePeriod = 0x33,
  GetEchoDecay = 0x34,
  SetEchoDecay = 0x35,
  MeasureOnce = 0x36,
  StartMeasurementLoop = 0x37,
  StopMeasurementLoop = 0x38,
}

export enum Sr04Rpt {
  Data = 0x91,
  SamplePeriod = 0x92,
  EchoDecay = 0x93,
}

/** echo_time_us sentinel: no echo received. */
export const ECHO_TIMEOUT = 0xffff;

export const SAMPLE_PERIOD_DEFAULT_US = 50_000;
export const ECHO_DECAY_DEFAULT_US = 5_000;
export const ECHO_DECAY_MIN_US = 4_000;
export const ECHO_DECAY_MAX_US = 65_000;

export interface Sr04Data {
  /** 0x36 single shot (host or SYNC_IN), 0x37 loop sample (ERRATA E3). */
  sourceCmd: number;
  timestampUs: bigint;
  echoTimeUs: number;
}

export function unpackSr04Data(p: Uint8Array): Sr04Data {
  const v = new DataView(p.buffer, p.byteOffset, p.byteLength);
  return {
    sourceCmd: v.getUint8(0),
    timestampUs: v.getBigUint64(1, true),
    echoTimeUs: v.getUint16(9, true),
  };
}

export function packSamplePeriod(periodUs: number): Uint8Array {
  const out = new Uint8Array(4);
  new DataView(out.buffer).setUint32(0, periodUs, true);
  return out;
}

export function unpackSamplePeriod(p: Uint8Array): number {
  return new DataView(p.buffer, p.byteOffset, p.byteLength).getUint32(0, true);
}

export function packEchoDecay(decayUs: number): Uint8Array {
  const out = new Uint8Array(2);
  new DataView(out.buffer).setUint16(0, decayUs, true);
  return out;
}

export function unpackEchoDecay(p: Uint8Array): number {
  return new DataView(p.buffer, p.byteOffset, p.byteLength).getUint16(0, true);
}

/**
 * Round-trip echo time → distance in mm; null for the timeout sentinel.
 * Default speed of sound 343 m/s; with airTempC uses c = 331.3 + 0.606·T.
 */
export function distanceMmFromEcho(echoTimeUs: number, airTempC?: number): number | null {
  if (echoTimeUs === ECHO_TIMEOUT) return null;
  const c = airTempC === undefined ? 343.0 : 331.3 + 0.606 * airTempC;
  return (echoTimeUs * c) / 2000.0;
}
